### DISCUSSION TITLE
Dynamic rule-based redirects (RFC)

### DISCUSSION THREAD

**Discussion by acelaya** [shlinkio/shlink#1912] [RFC]
I have been lately thinking on implementing an advanced system of rules to redirect visitors to different URLs based on a matching combination of conditions and a priority.

This would allow to add more dynamic redirects other than device redirects, with a way to handle condition conflicts and be able to predict where specific users will end-up being redirected.

This would work in combination with the required default long URL, which would always work as the redirect fallback for any users where no rules can be matched.

### Context

Historically, Shlink has only allowed every short URL to match a single long URL.

Some time ago, the first rule to dynamically redirect to different long URLs was introduced via [device-specific redirects](https://shlink.io/documentation/some-features/#device-specific-redirects), where you could define Android users to be redirected to a different place than iOS users.

This opened the door to new feature requests to support a similar thing for different languages, different locations, etc.

### Problem to solve

The main problem is that this would introduce a lot of complexity, and would require deciding what's the precedence of those rules.

For example, if an Android user whose system language is French visits a short URL which has a long URL for Android and another for French, which one has more precedence?

Because of that I have been more or less pushing back on those requests, but there's perhaps a way to redesign the whole system in a way that can be extended with more rules without breaking existing instances and in a predictable and flexible enough way.

### Proposed solution

The solution I'm thinking of is one in which certain rules are supported. These rules represent conditions that can only be checked at runtime, and every Shlink version will have a fixed set of supported rules.

Some examples could be:

* Query parameters with certain values.
* Language (via `Accept-Language` header).
* Device type (via `User-Agent` header).
* Location (via GeoLite2 geolocation).
* Domain (currently short URLs for a different domain are considered a different short URL, but it could be evolved to be defined as a rule).

When creating/editing a short URL, you would have to provide the default long URL, plus any number of URLs for any combination of rules.

It would also be required to define the priority for a combination of rules, so that in case of conflicting rules, Shlink knows which one to apply.

If none of the rules can be matched for a visitor, they would be redirected to the default long URL (which is Shlink's default behavior since forever).

This could allow a user to define a short URL which has these redirects:

1. Query param env=test AND language=fr-FR AND device=Android -> https://example.com/test-env
2. Device=iOS -> https://example.com/download-ios-app
3. Location=Spain -> https://example.com/hello-spaniards
4. Fallback (default long URL) -> https://example.com

At runtime, Shlink would check these rules and redirect to the long URL for the matching one.

### Challenges

Of course this is a relatively complex feature, and comes with a number of challenges to be solved. Some of them would be:

* Defining a good API model to represent rules, conditions, etc.
* Defining how rules and conditions are persisted.
* Querying rules with good performance, so that it does not impact redirecting visitors quickly.
* The UI to define and handle conditions. One thing to comes to mind is having a way to create the list of conditions for a short URL, and being able to drag'n'drop them in order to define the priority.
* Certain rules, like geolocation, can potentially have a lot of considerations. Should it match country? City? lat+long? A geofence?
* The feature should be backwards compatible, so device-specific URLs provided with current approach would need to be translated into the new rules system.
* The UI would need to work with different versions of Shlink, and therefore dynamically support device-specific redirects, rule-based redirects, or none.

**Comment by morphinestyle** [comment]
Wow!

It would be great to be able to give a language-based redirect!

Most of the URLs I currently have active lead to pages and downloads on multilingual sites, if you implement the language-based redirect I will finally be able to switch to Shlink and transfer all the existing URLs too!

(I'm already using a non-open source system with redirects based on language and other conditions, but would like to go open source with Shlink)

**Reply by acelaya** [reply]
Can I ask what service are you currently using? I would like to see how others solve this problem.

**Reply by morphinestyle** [reply]
Certain!

I happily use openmylink.io, using only language-based redirection individually.
But it could also set up multiple combined conditions (GEO, device, language), which I've never used.
If you need to do any tests let me know.

I also use Short.io with satisfaction, which allows redirects with multiple conditions, but I use the free version and I have no possibility of testing multiple conditions.

**Reply by acelaya** [reply]
So I guess I will have to add support to import links from those services into Shlink 😅

**Reply by morphinestyle** [reply]
Don't worry, I only have a few (but important) links that use language-based redirection: I'll do everything by hand, the important thing is that I can already configure the links before moving the domain pointing. 😁

Instead I have lots of links on Short.io, but unfortunately using the main domain shortcm.li, which I don't think I can move: I will have to reprint all the material little by little. 😭

They are both great services.
If I can help let me know.

**Reply by acelaya** [reply]
Just FYI, language-based redirects will be finally included with the overall rule-based functionality in Shlink 4.0.0. See [this comment](https://github.com/shlinkio/shlink/discussions/1912#discussioncomment-8582389) for more context.

**Comment by morphinestyle** [comment]
No update regarding Dynamic rule-based redirects?
Look forward! 😍

**Reply by acelaya** [reply]
It will still take time, as it requires deep changes.

I created this RFC mostly to gather feedback and describe how I would picture it at a high level.

These are some related issues, in case you want to watch them.

* https://github.com/shlinkio/shlink/issues/1914
* https://github.com/shlinkio/shlink/issues/1915

On top of the relative complexity, I'm currently focused on finishing shlink-web-client 4.0, and I won't start working on Shlink 4.0 until that's shipped.

**Comment by codaamok** [comment]
I personally do not have a need for this feature, but I can absolutely appreciate the difficulty in some of the challenges you've listed.

In particular with trying to create a good API to represent the rules. I work with other unrelated software which has a rule engine, and they do not let us configure rules via API but can only see a basic summary of them via `GET`. 

Perhaps consider this? At least initially, and add `POST` / `PATCH` / `DELETE` API support later - maybe people can only create rules via CLI to begin with. However, it does mean of course the UI will not be able to create the rules (at least initially), because I assume that uses the REST API as well.

I suggest this to help lower the barrier for the initial release of the feature.

Food for thought. Not perfect, but I'd say a good API for rule management is probably the hardest one in the list.

**Reply by acelaya** [reply]
I actually didn't think this would need to be supported by the CLI 😅

Regarding the API contract/endpoints, I also thought it might be reasonable to just expose GET and POST endpoints, and enforce all rules to be saved at once.

That way there's no need for DELETE or PATCH endpoints, at least for now.

**Comment by acelaya** [comment]
I have finally decided to include rules for [language-based](https://github.com/shlinkio/shlink/issues/1915) and [query-param-based](https://github.com/shlinkio/shlink/issues/1902) redirects in the initial implementation included with Shlink 4.0.0. The initial plan was to leave those for v4.1.0

The reason is that some specific rule is needed to properly test that the overall logic works as expected and performs well.

Also, once all the common logic to handle rules, conditions, etc. has been centralized, then the specific part for those two is surprisingly simple.

**Reply by morphinestyle** [reply]
WOW! I can't wait to test it! 😍
Are you creating a “system” capable of handling multiple conditions in the future? Did I get it right?

**Reply by acelaya** [reply]
> Are you creating a “system” capable of handling multiple conditions in the future? Did I get it right?

Yes, that's correct.

**Comment by acelaya** [comment]
The feature is mostly implemented, with initial support for language, query param and device type conditions.

Only a few details are missing, some benchmarking and finalizing the CLI interface is left.

The UI is also pending, but that's not part of Shlink per se.

Also, I finally decided to not keep backwards compatibility with device long urls, as it added a ton of complexity, and considering this goes in a new major version anyway, it was the time to change it.

Thanks to everyone who provided feedback.

**Reply by acelaya** [reply]
The feature was released yesterday with Shlink 4.0.0, and it is also now documented in the website https://shlink.io/documentation/advanced/dynamic-rule-based-redirects-system/

**Comment by morphinestyle** [comment]
WOW!
I will install it as soon as possible! 😍
